### PR TITLE
feat(network): get peer info

### LINKED ISSUE #564: Add command for querying info of connected peers
### Describe the feature

Add a new command+function for querying Peer info, mainly to provide

https://geth.ethereum.org/docs/rpc/ns-admin#admin_peers

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ayushm2003**
Closes #564 

Adds a new `PeerInfo` type and a new command `GetPeerInfo` to `NetworkHandleMessage`

### CODE REVIEW

**Comment by jonasbostoen** [comment]
Perhaps there should also be a command to get info about a certain peer by ID?

**Comment by mattsse** [comment]
> Perhaps there should also be a command to get info about a certain peer by ID?

good idea! let's add that

File: crates/net/network/src/network.rs [hunk 1]
```diff
@@ -211,4 +211,6 @@ pub(crate) enum NetworkHandleMessage {
     FetchClient(oneshot::Sender<FetchClient>),
     /// Apply a status update.
     StatusUpdate { height: u64, hash: H256, total_difficulty: U256 },
+	/// Get PeerInfo
+	GetPeerInfo,
```
Comment (mattsse) [inline crates/net/network/src/network.rs:215]: the way this should work is similar to `FetchClient` command. It should take a `oneshot::Sender` through which the `NetworManager` will send the response. The `Handle` can await an the receiver half of the channel and return the response asynchronously.

File: crates/net/network/src/session/handle.rs [hunk 2]
```diff
@@ -59,6 +63,21 @@ impl ActiveSessionHandle {
     }
 }
 
+#[derive(Debug)]
+#[allow(unused)]
+pub(crate) struct PeerInfo {
```
Comment (mattsse) [inline crates/net/network/src/session/handle.rs:68]: I think this should be public, right?

File: crates/net/network/src/session/mod.rs [hunk 3]
```diff
@@ -470,6 +470,22 @@ impl SessionManager {
             }
         }
     }
+
+	pub(crate) fn get_peer_info(&self) -> Vec<PeerInfo>{
+		let mut peers: Vec<PeerInfo> = Vec::new();
+
+		for (peer_id, session) in self.active_sessions.iter() {
```
Comment (mattsse) [inline crates/net/network/src/session/mod.rs:477]: nit: use `iter().map().collect()` instead of for loop and push to vec.

**Review by mattsse — CHANGES_REQUESTED** [review]
nice progress, left some early pointers

**Comment by ayushm2003** [comment]
> Perhaps there should also be a command to get info about a certain peer by ID?

Is it okay if I add them in the same pr?

File: crates/net/network/src/manager.rs [hunk 4]
```diff
@@ -476,6 +476,10 @@ where
                     self.swarm.state_mut().update_fork_id(transition.current);
                 }
             }
+			NetworkHandleMessage::GetPeerInfo => {
+				let peer_info = self.swarm.sessions_mut().get_peer_info();
+				println!("{:#?}", peer_info);
```
Comment (onbjerg) [inline crates/net/network/src/manager.rs:481]: nit

```suggestion
```

**Review by onbjerg — CHANGES_REQUESTED** [review]

**Comment by mattsse** [comment]
> > Perhaps there should also be a command to get info about a certain peer by ID?
> 
> Is it okay if I add them in the same pr?

sure!

**Comment by ayushm2003** [comment]
Hey, I had a couple of questions -

1. Can we add the field `ClientId` to the [`Established`](https://github.com/paradigmxyz/reth/blob/c9958a4d13be69977032e306021de71584d08757/crates/net/network/src/session/handle.rs#LL70) event? The [doc](https://github.com/ethereum/devp2p/blob/6b0abc3d956a626c28dce1307ee9f546db17b6bd/rlpx.md#hello-0x00) mentioned in the comment does specify it as one of the parameters. This would allow us to add the field to [`ActiveSessionHandle`](https://github.com/paradigmxyz/reth/blob/c9958a4d13be69977032e306021de71584d08757/crates/net/network/src/session/mod.rs#L376)
2. I am not sure if there is a way to test the command manually, if so how?

File: crates/net/network/src/manager.rs [hunk 5]
```diff
@@ -434,7 +434,7 @@ where
     }
 
     /// Handler for received messages from a handle
-    fn on_handle_message(&mut self, msg: NetworkHandleMessage) {
+    async fn on_handle_message(&mut self, msg: NetworkHandleMessage) {
```
Comment (mattsse) [inline crates/net/network/src/manager.rs:437]: we don't need async here

**Review by mattsse — CHANGES_REQUESTED** [review]
we don't need async code in the manager.

for reference, see `NetworkHandleMessage::FetchClient`

**Comment by mattsse** [comment]
> an we add the field ClientId to the Established event.

yes let's do that

**Comment by mattsse** [comment]
> I am not sure if there is a way to test the command manually, if so how?

See this test for reference which connects peers manually. for the test we want the same (2 peers are enough) and then retrieve the info once they're connected.

https://github.com/paradigmxyz/reth/blob/main/crates/net/network/tests/it/connect.rs#L92

**Comment by gakonst** [comment]
Hi @ayushm2003 let us know on your plans for this PR or if we should take it over^^

**Comment by ayushm2003** [comment]
Hey, Sorry for the delay. I have a commit ready to fix the compilation errors and will add the test later today. I don't really understand the [async requirement](https://github.com/paradigmxyz/reth/pull/603#discussion_r1056826691), so would appreciate some help in the form of a little more explanation or a commit. 

Would this be ok?

**Comment by mattsse** [comment]
sending does not require async for `tokio::oneshot::Sender`, we don't want any `async/await` inside the control loop of the Network, we do however want async/await on the `NetworkHandle` which awaits the response on the `tokio::oneshot::Receiver`

**Review by mattsse — APPROVED** [review]
nice job!

tysm

**Pull request merged**